CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxBaseNodesFile.swift

```diff
@@ -319,7 +214,85 @@ let syntaxBaseNodesFile = SourceFileSyntax(leadingTrivia: copyrightHeader) {
   }
 
   leafProtocolDecl(type: "_LeafSyntaxNodeProtocol", inheritedType: "SyntaxProtocol")
-  leafProtocolExtension(type: "_LeafSyntaxNodeProtocol", inheritedType: "SyntaxProtocol")
+  leafImpossibleCastsExtension(type: "_LeafSyntaxNodeProtocol", inheritedType: "SyntaxProtocol")
+}
+
+private func upcastExtension(for node: Node) -> DeclSyntax {
+  return DeclSyntax(
+    #"""
+    /// Extension of ``\#(node.kind.protocolType)`` to provide overloads of the casting
+    /// methods for upcasting, where they are known to always succeed.
+    extension \#(node.kind.protocolType) {
+      /// Checks if the current syntax node can be upcast to its base node type (``\#(node.kind.syntaxType)``).
+      ///
+      /// - Returns: `true` since the node can always be upcast to its base node.
+      @available(*, deprecated, message: "This cast will always succeed")
+      public func `is`(_ syntaxType: \#(node.kind.syntaxType).Type) -> Bool {
+        return true
+      }
+
+      /// Attempts to upcast the current syntax node to its base node type (``\#(node.kind.syntaxType)``).
+      ///
+      /// - Returns: The base node created from the current syntax node, as the node can
+      ///            always be upcast to its base type.
+      @available(*, deprecated, message: "Use `\#(node.kind.syntaxType).init` for upcasting")
+      public func `as`(_ syntaxType: \#(node.kind.syntaxType).Type) -> \#(node.kind.syntaxType)? {
+        return \#(node.kind.syntaxType)(self)
+      }
+
+      /// Force-upcast the current syntax node to its base node type (``\#(node.kind.syntaxType)``).
+      ///
+      /// - Returns: The base node created from the current syntax node, as the node can
+      ///            always be upcast to its base type.
+      @available(*, deprecated, message: "Use `\#(node.kind.syntaxType).init` for upcasting")
+      public func cast(_ syntaxType: \#(node.kind.syntaxType).Type) -> \#(node.kind.syntaxType) {
+        return \#(node.kind.syntaxType)(self)
+      }
```

Comment:
FWIW I'm not sure we should prefer the constructor here. If it wasn't for the performance, we'd use existential's instead and then upcasting with the regular `as` would be the way to do this. `.as` is analogous so preferring the init is a little odd. Have left it in for now though.

Reply:
> In my understating it distinguishes between using constructors and as methods for upcasting to base node types

Thanks for pointing that out! If we didn't need `ExprSyntax` then the cast wouldn't be needed anyway, so it probably doesn't matter what we choose. Given we have docs suggesting the constructor already, I'll probably just leave it.